File: Baekjoon/분할 정복/MinjuCho/1992.cpp
// 문제 풀이 링크: https://velog.io/@cuppizza/백준-1992-쿼드트리-파이썬-C
#include <vector>
#include <stdio.h>

using namespace std;

vector<vector<char>> paper;
vector<char> result;
void div_half(int x, int y, int n)
{
	if (n == 1)
	{
		result.push_back(paper[y][x]);
		return;
	}
	else
	{
		char pivot = paper[y][x];
		bool isPivot = true;
		for (int i = y; i < y + n; ++i)
		{
			for (int j = x; j < x + n; ++j)
			{
				if (paper[i][j] != pivot)
				{
					isPivot = false;
					break;
				}
			}
			if (!isPivot)
			{
				break;
			}
		}
		if (!isPivot)
		{
			result.push_back('(');
			div_half(x, y, n / 2);
			div_half(x + n / 2, y, n / 2);
			div_half(x, y + n / 2, n / 2);
			div_half(x + n / 2, y + n / 2, n / 2);
			result.push_back(')');
		}
		else
		{
			result.push_back(paper[y][x]);
			return;
		}
	}
}


int main(void)
{
	int N;
	scanf_s("%d", &N);

	for (int i = 0; i < N; ++i)
	{
		vector<char> row(N);
		for (int j = 0; j < N;)
		{
			char c;
			scanf_s("%c", &c);
			if (c != '\n' && c != '\0')
			{
				row[j++] = c;
			}
		}
		paper.push_back(row);
	}

	div_half(0, 0, N);
	for (char c : result)
	{
		printf("%c", c);
	}
}

File: Baekjoon/분할 정복/MinjuCho/1992.py
# 문제 풀이 링크: https://velog.io/@cuppizza/백준-1992-쿼드트리-파이썬-C
import sys
input = sys.stdin.readline
N = int(input())
paper = []
for _ in range(N):
    paper.append(list(input().rstrip()))
result = []

def div_half(x, y, n):
    half = int(n/2)
    if n == 1:
        result.append(paper[y][x])
        return
    else:
        pivot = paper[y][x]
        isPivot = True
        for i in range(y, y+n):
            for j in range(x, x+n):
                if paper[i][j] != pivot:
                    isPivot = False
                    break
            if isPivot == False:
                break
        if isPivot == False:
            result.append("(")
            div_half(x, y, half)
            div_half(x+half, y, half)
            div_half(x, y+half, half)
            div_half(x+half, y+half, half)
            result.append(")")
        else:
            result.append(paper[y][x])
            return

div_half(0,0,N)
result = ''.join(result)
print(result)
